Cross-compilation of ARM bare-metal samples broken due to #7721

After updating the IREE submodule in https://github.com/iml130/iree-bare-metal-arm to https://github.com/google/iree/commit/658cf6c8a48d155259f3bb08a536cc47f1d50b78 or newer, the static library example is broken.
When cross-compiling the static library example for ARMv7em, the linkage fails with the following error:

```
[110/110] Linking C executable samples/sample_static_library
FAILED: samples/sample_static_library 
: && /usr/local/gcc-arm-none-eabi-10-2020-q4-major/bin/arm-none-eabi-gcc -mthumb -march=armv7e-m -mfloat-abi=hard -mfpu=fpv4-sp-d16 -DIREE_TIME_NOW_FN="{ return 0; }"   -O3 -DNDEBUG -lc -lm -lnosys -T /home/admin-mbrehler/repos/iree-bare-metal-arm/build_tools/stm32f407-cmsis.ld  -specs=nosys.specs    -lm samples/CMakeFiles/sample_static_library.dir/static_library_demo.c.obj -o samples/sample_static_library -L/home/admin-mbrehler/repos/iree-bare-metal-arm/build_tools samples/libsimple_mul_c.a  third_party/iree/iree/hal/local/loaders/libiree_hal_local_loaders_static_library_loader.a  third_party/iree/iree/hal/local/libiree_hal_local_sync_driver.a  samples/libsimple_mul.a  build_tools/third_party/cmsis_device_f4/libcmsis_device_f4.a  third_party/iree/iree/runtime/libiree_runtime_impl.a  third_party/iree/iree/base/internal/libiree_base_internal_file_io.a  third_party/iree/iree/hal/drivers/libiree_hal_drivers_drivers.a  third_party/iree/iree/hal/dylib/registration/libiree_hal_dylib_registration_sync.a  third_party/iree/iree/hal/local/loaders/libiree_hal_local_loaders_embedded_library_loader.a  third_party/iree/iree/hal/local/elf/libiree_hal_local_elf_elf_module.a  third_party/iree/iree/hal/local/elf/libiree_hal_local_elf_arch.a  third_party/iree/iree/hal/local/elf/libiree_hal_local_elf_platform.a  third_party/iree/iree/hal/vmvx/registration/libiree_hal_vmvx_registration_sync.a  third_party/iree/iree/hal/local/libiree_hal_local_sync_driver.a  third_party/iree/iree/base/internal/libiree_base_internal_arena.a  third_party/iree/iree/base/internal/libiree_base_internal_atomic_slist.a  third_party/iree/iree/hal/local/loaders/libiree_hal_local_loaders_vmvx_module_loader.a  third_party/iree/iree/modules/vmvx/libiree_modules_vmvx_vmvx.a  third_party/iree/iree/modules/hal/libiree_modules_hal_hal.a  third_party/iree/iree/vm/libiree_vm_bytecode_module.a  third_party/iree/iree/vm/libiree_vm_impl.a  third_party/iree/build_tools/third_party/flatcc/libflatcc_parsing.a  -lm  third_party/iree/iree/hal/local/libiree_hal_local_local.a  third_party/iree/iree/hal/libiree_hal_hal.a  third_party/iree/iree/base/internal/libiree_base_internal_synchronization.a  third_party/iree/iree/base/internal/libiree_base_internal_fpu_state.a  third_party/iree/iree/base/libiree_base_base.a && :
/usr/local/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /usr/local/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libm.a(lib_a-math_errf.o): in function `__math_oflowf':
math_errf.c:(.text.__math_oflowf+0x0): multiple definition of `__math_oflowf'; samples/libsimple_mul.a(simple_mul.o):simple_mul_dispatch_0:(.text.__math_oflowf+0x0): first defined here
/usr/local/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /usr/local/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libm.a(lib_a-math_errf.o): in function `__math_invalidf':
math_errf.c:(.text.__math_invalidf+0x0): multiple definition of `__math_invalidf'; samples/libsimple_mul.a(simple_mul.o):simple_mul_dispatch_0:(.text.__math_invalidf+0x0): first defined here
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```



**To Reproduce**
Steps to reproduce the behavior:

Prior to compile and install the host build execute:

```bash
cd third_party/iree
git pull
git checkout 658cf6c8a48d155259f3bb08a536cc47f1d50b78
git submodule update
cd ../../
```
Follow the instructions in https://github.com/iml130/iree-bare-metal-arm#host-build for the host build afterwards.
Following the description in https://github.com/iml130/iree-bare-metal-arm#target-build and building
```bash
cmake --build . --target sample_static_library
```
fails with the error posted above. Any help is appreciated!


**Side note**
Building the embedded dynamic library example via
```bash
cmake --build . --target sample_embedded_sync 
```
now fails with a much more expressive error message which is great :D This is expected as `__aeabi_fmul`is defined in libgcc.
```
FAILED: samples/simple_embedding_test_module_dylib_arm_32.vmfb 
cd /home/admin-mbrehler/repos/iree-bare-metal-arm/build-cmsis/samples && /home/admin-mbrehler/repos/iree-bare-metal-arm/build-iree-host-install/bin/iree-translate -iree-input-type=mhlo -iree-mlir-to-vm-bytecode-module -iree-hal-target-backends=dylib-llvm-aot -iree-llvm-target-triple=armv7em-pc-linux-elf -iree-llvm-target-float-abi=hard -iree-llvm-debug-symbols=false -iree-vm-bytecode-module-strip-source-map=true -iree-vm-emit-polyglot-zip=false /home/admin-mbrehler/repos/iree-bare-metal-arm/third_party/iree/iree/samples/simple_embedding/simple_embedding_test.mlir -o simple_embedding_test_module_dylib_arm_32.vmfb
lld: error: undefined symbol: __aeabi_fmul
>>> referenced by simple_embedding_test.mlir:3 (/home/admin-mbrehler/repos/iree-bare-metal-arm/third_party/iree/iree/samples/simple_embedding/simple_embedding_test.mlir:3)
>>>               /tmp/simple_mul_dispatch_0-7886fe.o:(.Lsimple_mul_dispatch_0)
Linking failed; escaped command line returned exit code 256:
```

--- a/iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp
+++ b/iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp
@@ -103,10 +103,11 @@ static LogicalResult appendDebugDatabase(std::vector<int8_t> &baseFile,
 // Verifies builtin bitcode is loaded correctly and appends it to |linker|.
 //
 // Example:
-//  if (failed(linkBuiltinLibrary(loc, linker, "libfoo", loadLibFoo(...))))
+//  if (failed(linkBuiltinLibrary(loc, linker, linkerFlag, targetMachine,
+//  "libfoo", loadLibFoo(...))))
 static LogicalResult linkBuiltinLibrary(
-    Location loc, llvm::Linker &linker, llvm::TargetMachine *targetMachine,
-    StringRef name,
+    Location loc, llvm::Linker &linker, llvm::Linker::Flags linkerFlag,
+    llvm::TargetMachine *targetMachine, StringRef name,
     llvm::Expected<std::unique_ptr<llvm::Module>> bitcodeModuleValue) {
   // Ensure the bitcode loaded correctly. It may fail if the LLVM version is
   // incompatible.
@@ -122,9 +123,7 @@ static LogicalResult linkBuiltinLibrary(
 
   // Link the bitcode into the base module. This will merge in any required
   // symbols and override declarations that may exist.
-  if (linker.linkInModule(
-          std::move(bitcodeModule),
-          llvm::Linker::OverrideFromSrc /*| llvm::Linker::LinkOnlyNeeded*/)) {
+  if (linker.linkInModule(std::move(bitcodeModule), linkerFlag)) {
     return mlir::emitError(loc) << "failed to link " << name << " bitcode";
   }
 
@@ -340,16 +339,20 @@ class LLVMAOTTargetBackend final : public TargetBackend {
     // Note that if producing a static library then the symbols we add must be
     // weak such that we don't trigger ODR issues.
     llvm::Linker moduleLinker(*llvmModule);
+
+    llvm::Linker::Flags linkerFlag = llvm::Linker::OverrideFromSrc;
+    if (options_.linkStatic) linkerFlag = llvm::Linker::LinkOnlyNeeded;
+
     if (failed(linkBuiltinLibrary(
-            variantOp.getLoc(), moduleLinker, targetMachine.get(), "libdevice",
-            loadDeviceBitcode(targetMachine.get(), context)))) {
+            variantOp.getLoc(), moduleLinker, linkerFlag, targetMachine.get(),
+            "libdevice", loadDeviceBitcode(targetMachine.get(), context)))) {
       return mlir::emitError(variantOp.getLoc())
              << "failed linking in builtin library for target triple '"
              << options_.targetTriple << "'";
     }
     if (failed(linkBuiltinLibrary(
-            variantOp.getLoc(), moduleLinker, targetMachine.get(), "libmusl",
-            loadMuslBitcode(targetMachine.get(), context)))) {
+            variantOp.getLoc(), moduleLinker, linkerFlag, targetMachine.get(),
+            "libmusl", loadMuslBitcode(targetMachine.get(), context)))) {
       return mlir::emitError(variantOp.getLoc())
              << "failed linking in builtin library for target triple '"
              << options_.targetTriple << "'";

